File: frontend/src/lib/api/archive.ts
/**
 * Public entry-point for the archive API. Mirrors the backend contract in
 * `backend/app/api/archive.py` + `backend/app/models/archive.py`.
 *
 * Two implementations are wired behind the same {@link ArchiveApi} surface:
 *
 *  - **Default everywhere**: real HTTP client against `/api/v1/archive/*`
 *    ({@link archive.client.ts}).
 *  - **Dev with `VITE_USE_MOCK_ARCHIVE_API=true`**: localStorage-backed mock
 *    shim ({@link archive.mock.ts}). Opt-in only — used for offline work
 *    when the backend isn't reachable. Production builds ignore the flag
 *    and always use the real client.
 *
 * Backend specifics:
 *  - Archive identity is `(network, tx_hash)` — every call carries `network`.
 *  - Notes live in a single free-text `note` field. The Delete dialog in the
 *    Detail page composes its "reason" + free notes into this one string.
 *  - Bulk import is **skip-existing**: existing rows are never overwritten
 *    (response shape has `inserted`/`skipped`, no `updated`).
 *  - Export CSV is **server-side**: callers fetch it via {@link ArchiveApi.download}
 *    and hand the Blob to a programmatic anchor click — there's no client
 *    paginated fetch.
 */
import { archiveApiClient } from "./archive.client";
import { archiveApiMock } from "./archive.mock";

import type { Severity } from "@/lib/attacks";

/* ---------- Wire format ---------- */

/** Cardano network identifier. Same enum as the backend `NetworkType`. */
export type Network = "mainnet" | "preprod" | "preview";

/** Backend response row for `GET /api/v1/archive`. */
export type ArchiveEntry = {
	network: Network;
	tx_hash: string;
	/** Free-text "reason + notes" composed by the UI on archive. */
	note: string;
	archived_by: string;
	/** ISO datetime, set by the server. */
	archived_at: string;
	/** Origin tag: `"local"` or `"import:<source_label>"`. */
	source: string;
	/* ---- joined from tx_class_scores (null when archive came from CSV import
	   for a tx this instance has never observed locally) ---- */
	max_score: number | null;
	max_class: string | null;
	risk_band: string | null;
	analyzed_at: string | null;
};

/** Payload for `POST /api/v1/archive`. */
export type ArchiveCreateRequest = {
	network: Network;
	tx_hash: string;
	note: string;
	archived_by: string;
};

/** Entry inside a bulk-import body. */
export type ArchiveBulkEntry = ArchiveCreateRequest & {
	/** Original archive timestamp from the source instance, if known. */
	archived_at?: string;
	/** Original `source` tag from the exported CSV (ignored, kept for round-trip). */
	source?: string;
};

export type ArchiveListParams = {
	network?: Network;
	/** Inclusive lower bound on `archived_at` (ISO datetime). */
	from?: string;
	/** Inclusive upper bound on `archived_at` (ISO datetime). */
	to?: string;
	limit?: number;
	offset?: number;
};

export type ArchiveListResponse = {
	count: number;
	total: number;
	data: ArchiveEntry[];
};

export type ArchiveBulkResponse = {
	inserted: number;
	skipped: number;
	errors: string[];
};

export type ArchiveExportParams = {
	network?: Network;
	from?: string;
	to?: string;
};

/* ---------- Public contract ---------- */

export interface ArchiveApi {
	/** Paginated list. Date range bounds are inclusive on both ends. */
	list(params: ArchiveListParams): Promise<ArchiveListResponse>;

	/** Single entry by tx_hash, `null` if not archived. */
	get(txHash: string, network?: Network): Promise<ArchiveEntry | null>;

	/** Archive one alert. Backend returns 409 on duplicate `(network, tx_hash)`. */
	create(entry: ArchiveCreateRequest): Promise<void>;

	/** Restore (hard-delete) the archive row. 204/404 both map to `void`. */
	remove(txHash: string, network?: Network): Promise<void>;

	/**
	 * Bulk-upsert with **skip-existing** semantics (never overwrites).
	 * `sourceLabel` tags the origin instance on inserted rows.
	 */
	bulk(
		entries: ArchiveBulkEntry[],
		sourceLabel: string,
	): Promise<ArchiveBulkResponse>;

	/**
	 * Download the CSV for the current params and return it as a Blob plus
	 * the suggested filename (parsed from the server's Content-Disposition).
	 *
	 * Goes through `fetchWithAuth` (session cookie) rather than a plain
	 * `<a download href>` so an expired session surfaces as a typed 401
	 * (UnauthorizedError → login redirect) instead of a silently failed
	 * navigation, and so the server's Content-Disposition filename can be
	 * parsed. The page wires this blob to a programmatic anchor click to
	 * trigger the browser download.
	 */
	download(params: ArchiveExportParams): Promise<{
		blob: Blob;
		filename: string;
	}>;
}

/* ---------- Implementation selection ---------- */

// Real backend by default. The mock shim is opt-in via env var and only
// honored in dev builds — production always talks to the real backend so a
// stray env var can't silently switch a deployment to localStorage.
const useMock =
	import.meta.env.DEV &&
	import.meta.env.VITE_USE_MOCK_ARCHIVE_API === "true";

export const archiveApi: ArchiveApi = useMock
	? archiveApiMock
	: archiveApiClient;

export const isUsingMockArchive = useMock;

/* ---------- Type re-exports for downstream consumers ---------- */

export type { Severity };

File: backend/app/models/archive.py
"""Archive models: mark flagged transactions as known false positives.

The archive subsystem is additive: archived rows live in their own
``archived_alerts`` ClickHouse table and never mutate ``tx_class_scores``.
Cross-instance CSV import is supported, so an entry may exist for a
transaction this instance has never observed (no matching score row).
"""

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.transaction import NetworkType

# 64-char lowercase hex. Cardano tx hashes are blake2b-256 of the tx body.
TX_HASH_PATTERN = r"^[0-9a-f]{64}$"


class ArchiveEntry(BaseModel):
    """Input shape for POST /api/archive and entries inside bulk import."""

    network: NetworkType
    tx_hash: str = Field(pattern=TX_HASH_PATTERN)
    note: str = Field(min_length=1, max_length=2000)
    archived_by: str = Field(min_length=1, max_length=200)


class ArchiveBulkEntry(ArchiveEntry):
    """Bulk import entry. Carries the original archived_at + source so a
    CSV exported from one instance round-trips cleanly into another."""

    archived_at: datetime | None = None
    source: str | None = None


class BulkArchiveRequest(BaseModel):
    # Cap kept well below the ClickHouse default max_query_size (~256 KiB).
    # The existence check sends one tuple per entry in a single IN clause;
    # 2000 entries at ~80 bytes per (network, tx_hash) tuple stays around
    # 160 KiB worst case, leaving headroom for the surrounding SQL.
    entries: list[ArchiveBulkEntry] = Field(min_length=1, max_length=2_000)
    source_label: str = Field(
        min_length=1,
        max_length=200,
        description=(
            "Identifier of the originating instance, recorded on inserted "
            "rows as 'import:<source_label>'. Local archives use 'local'."
        ),
    )


class BulkArchiveResult(BaseModel):
    inserted: int
    skipped: int
    errors: list[str] = Field(default_factory=list)


class ArchiveEntryEnriched(BaseModel):
    """GET /api/archive response row: archive metadata + the original
    detection record (when present locally) joined from tx_class_scores."""

    network: str
    tx_hash: str
    note: str
    archived_by: str
    archived_at: datetime
    source: str
    # Joined detection data; null when this archive came from a CSV import
    # for a tx this instance never observed locally.
    max_score: float | None = None
    max_class: str | None = None
    risk_band: str | None = None
    analyzed_at: datetime | None = None
